File: src/test/java/no/nav/fo/veilarbregistrering/arbeidsforhold/adapter/ArbeidsforholdGatewayTest.java
package no.nav.fo.veilarbregistrering.arbeidsforhold.adapter;

import no.nav.fo.veilarbregistrering.arbeidsforhold.ArbeidsforholdGateway;
import no.nav.fo.veilarbregistrering.bruker.Foedselsnummer;
import no.nav.fo.veilarbregistrering.config.CacheConfig;
import org.junit.jupiter.api.AfterAll;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.config.BeanDefinition;
import org.springframework.beans.factory.support.BeanDefinitionBuilder;
import org.springframework.context.annotation.AnnotationConfigApplicationContext;

import java.util.Arrays;
import java.util.Collections;

import static org.mockito.Mockito.*;

class ArbeidsforholdGatewayTest {

    private static AnnotationConfigApplicationContext context;

    private static AaregRestClient aaregRestClient;

    @BeforeAll
    public static void setup() {
        aaregRestClient = mock(AaregRestClient.class);

        BeanDefinition beanDefinition = BeanDefinitionBuilder.rootBeanDefinition(ArbeidsforholdGatewayImpl.class)
                .addConstructorArgValue(aaregRestClient).getBeanDefinition();

        context = new AnnotationConfigApplicationContext();
        context.register(CacheConfig.class);
        context.getDefaultListableBeanFactory().registerBeanDefinition("arbeidsforhold", beanDefinition);
        context.refresh();
        context.start();
    }

    @AfterAll
    public static void tearDown() {
        context.stop();
    }

    @Test
    public void skalCacheVedKallPaaSammeFnr() {
        ArbeidsforholdGateway arbeidsforholdGateway = context.getBean(ArbeidsforholdGateway.class);
        when(aaregRestClient.finnArbeidsforhold(any())).thenReturn(Arrays.asList(getArbeidsforholdDto()));
        arbeidsforholdGateway.hentArbeidsforhold(Foedselsnummer.of("12345678910"));
        arbeidsforholdGateway.hentArbeidsforhold(Foedselsnummer.of("12345678910"));
        verify(aaregRestClient, times(1)).finnArbeidsforhold(any());
    }

    @Test
    public void skalIkkeCacheVedKallPaaForskjelligFnr() {
        ArbeidsforholdGateway arbeidsforholdGateway = context.getBean(ArbeidsforholdGateway.class);
        when(aaregRestClient.finnArbeidsforhold(any())).thenReturn(Arrays.asList(getArbeidsforholdDto()));
        arbeidsforholdGateway.hentArbeidsforhold(Foedselsnummer.of("12345678910"));
        arbeidsforholdGateway.hentArbeidsforhold(Foedselsnummer.of("109987654321"));
        verify(aaregRestClient, times(2)).finnArbeidsforhold(any());
    }

    private ArbeidsforholdDto getArbeidsforholdDto() {

        ArbeidsforholdDto arbeidsforholdDto = new ArbeidsforholdDto();

        ArbeidsgiverDto arbeidsgiverDto = new ArbeidsgiverDto();
        arbeidsgiverDto.setOrganisasjonsnummer("981129687");
        arbeidsgiverDto.setType("Organisasjon");
        arbeidsforholdDto.setArbeidsgiver(arbeidsgiverDto);

        AnsettelsesperiodeDto ansettelsesperiodeDto = new AnsettelsesperiodeDto();
        PeriodeDto periodeDto = new PeriodeDto();
        periodeDto.setFom("2014-07-01");
        periodeDto.setTom("2015-12-31");
        ansettelsesperiodeDto.setPeriode(periodeDto);
        arbeidsforholdDto.setAnsettelsesperiode(ansettelsesperiodeDto);

        ArbeidsavtaleDto arbeidsavtaleDto = new ArbeidsavtaleDto();
        arbeidsavtaleDto.setYrke("2130123");
        arbeidsforholdDto.setArbeidsavtaler(Collections.singletonList(arbeidsavtaleDto));

        arbeidsforholdDto.setNavArbeidsforholdId(123456);

        return arbeidsforholdDto;
    }
}

File: src/test/java/no/nav/fo/veilarbregistrering/arbeidsforhold/adapter/ArbeidsforholdGatewayTest.kt
package no.nav.fo.veilarbregistrering.arbeidsforhold.adapter

import io.mockk.every
import io.mockk.mockk
import io.mockk.verify
import no.nav.fo.veilarbregistrering.arbeidsforhold.ArbeidsforholdGateway
import no.nav.fo.veilarbregistrering.bruker.Foedselsnummer
import no.nav.fo.veilarbregistrering.config.CacheConfig
import org.junit.jupiter.api.*
import org.springframework.beans.factory.config.BeanDefinition
import org.springframework.beans.factory.support.BeanDefinitionBuilder
import org.springframework.context.annotation.AnnotationConfigApplicationContext

class ArbeidsforholdGatewayTest {

    @BeforeEach
    fun setup() {
        aaregRestClient = mockk(relaxed = true)
        val beanDefinition: BeanDefinition =
            BeanDefinitionBuilder.rootBeanDefinition(ArbeidsforholdGatewayImpl::class.java)
                .addConstructorArgValue(aaregRestClient).beanDefinition
        context = AnnotationConfigApplicationContext()
        context.register(CacheConfig::class.java)
        context.defaultListableBeanFactory.registerBeanDefinition("arbeidsforhold", beanDefinition)
        context.refresh()
        context.start()
    }

    @AfterEach
    fun tearDown() {
        context.stop()
    }

    @Test
    fun skalCacheVedKallPaaSammeFnr() {
        val arbeidsforholdGateway = context.getBean(
            ArbeidsforholdGateway::class.java
        )
        every { aaregRestClient.finnArbeidsforhold(any()) } returns emptyList()
        arbeidsforholdGateway.hentArbeidsforhold(IDENT_1)
        arbeidsforholdGateway.hentArbeidsforhold(IDENT_1)
        verify(exactly = 1) { aaregRestClient.finnArbeidsforhold(any()) }
    }

    @Test
    fun skalIkkeCacheVedKallPaaForskjelligFnr() {
        val arbeidsforholdGateway = context.getBean(
            ArbeidsforholdGateway::class.java
        )
        every { aaregRestClient.finnArbeidsforhold(any()) } returns emptyList()
        arbeidsforholdGateway.hentArbeidsforhold(IDENT_1)
        arbeidsforholdGateway.hentArbeidsforhold(IDENT_2)

        verify(exactly = 1) { aaregRestClient.finnArbeidsforhold(IDENT_2) }
    }

    companion object {
        private lateinit var aaregRestClient: AaregRestClient
        private lateinit var context: AnnotationConfigApplicationContext
        private val IDENT_1 = Foedselsnummer.of("12345678910")
        private val IDENT_2 = Foedselsnummer.of("109987654321")
    }
}